# Assignment 10

## Watch, Learn, and Practice

Watch the following video on Youtube (about 1 hour 30 minutes long):

https://www.youtube.com/watch?v=iYie42M1ZyU 

This video discusses the concept of **tidy data** and how to use Pandas to make data tidy. 

This is a great refresher for Pandas data analysis. You will also learn new things that we have not covered in our class. 

The data files and notebooks used in the video are hosted in GitHub: 

https://github.com/chendaniely/pydatadc_2018-tidy

## Instructions:
- You must download data files from GitHub prior to playing the video.
- You must create/write/run your own notebook as you follow along with the speaker. 
    - You may download the notebooks from the above GitHub repo and copy/paste cells into yours to save typing time
- You will pause and work on your notebook and continue as you ride along.
- After you are done, create a **markdown cell** at the end of your notebook and write **a list of new things** you learned from this video.
- You will upload your Jupyter notebook file to GitHub and submit the link in BlackBoard. 


# Introduction

This notebook basically focuses on learning and working on data tyding using the pandas library in python. We'll look at the various methods used in python to tidy data which are very useful in various application based projects. 

Moreover, he github link and the access to the files that will be discussed in this notebook is given below. Please make sure to download these files before moving on further. 

"https://github.com/chendaniely/pydatadc_2018-tidy"



##Pandas

Pandas is one of the many libraries that python consists of. Due to the large amount of data that it deals with, we will have to import a particular library(in this case pandas) to work with the functions and methods in it.

Pandas is used to deal with tabular data. Basically it allows us to read data, manipulate data and also delete the data in it, with simple lines of code.

To get started with, we will work on few basic and most important commands in pandas to read a file and extract the required data from it.

In [1]:
import pandas

df = pandas.read_csv('gapminder.tsv', sep = '\t') #reads the data file 

In [2]:
df.head() #shows the first 5 columns and rows od the tablular data

country continent  year  lifeExp       pop   gdpPercap
0  Afghanistan      Asia  1952   28.801   8425333  779.445314
1  Afghanistan      Asia  1957   30.332   9240934  820.853030
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
3  Afghanistan      Asia  1967   34.020  11537966  836.197138
4  Afghanistan      Asia  1972   36.088  13079460  739.981106

In [3]:
df.columns #shows all the column names in the dataframe

Index(['country', 'continent', 'year', 'lifeExp', 'pop', 'gdpPercap'], dtype='object')

In [4]:
df.index #shows us where the dataframe starts and ends 

RangeIndex(start=0, stop=1704, step=1)

In [5]:
df.values #shows us the body of the dataframe

array([['Afghanistan', 'Asia', 1952, 28.801, 8425333, 779.4453145],
       ['Afghanistan', 'Asia', 1957, 30.331999999999997, 9240934,
        820.8530296],
       ['Afghanistan', 'Asia', 1962, 31.997, 10267083, 853.1007099999999],
       ...,
       ['Zimbabwe', 'Africa', 1997, 46.809, 11404948, 792.4499602999999],
       ['Zimbabwe', 'Africa', 2002, 39.989000000000004, 11926563,
        672.0386227000001],
       ['Zimbabwe', 'Africa', 2007, 43.486999999999995, 12311143,
        469.70929810000007]], dtype=object)

In [6]:
type(df) #shows us the type of our dataframe

pandas.core.frame.DataFrame

In [7]:
df.shape  #gives us the number of rows and columns in our dataframe

(1704, 6)

In [8]:
df.info() #gives the structure of the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   country    1704 non-null   object 
 1   continent  1704 non-null   object 
 2   year       1704 non-null   int64  
 3   lifeExp    1704 non-null   float64
 4   pop        1704 non-null   int64  
 5   gdpPercap  1704 non-null   float64
dtypes: float64(2), int64(2), object(2)
memory usage: 80.0+ KB


In [9]:
df['country'] #to extract a sepecific column from our dataframe

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
1699       Zimbabwe
1700       Zimbabwe
1701       Zimbabwe
1702       Zimbabwe
1703       Zimbabwe
Name: country, Length: 1704, dtype: object

## Subsetting Columns of a Dataframe

In [10]:
subset = df[['country','continent','year']] #to extract multiple columns from a dataframe into a another single dataframe

subset.head()

country continent  year
0  Afghanistan      Asia  1952
1  Afghanistan      Asia  1957
2  Afghanistan      Asia  1962
3  Afghanistan      Asia  1967
4  Afghanistan      Asia  1972

## Subsetting Rows or Index of a Dataframe

In [11]:
df.loc[2] #gives us the row with index number 2

country      Afghanistan
continent           Asia
year                1962
lifeExp           31.997
pop             10267083
gdpPercap        853.101
Name: 2, dtype: object

In [12]:
df.loc[[2,0]] #gives us multiple required rows from the dataframe

country continent  year  lifeExp       pop   gdpPercap
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
0  Afghanistan      Asia  1952   28.801   8425333  779.445314

In [13]:
df.iloc[2] #gives us the data of the 2nd Row of the dataframe

country      Afghanistan
continent           Asia
year                1962
lifeExp           31.997
pop             10267083
gdpPercap        853.101
Name: 2, dtype: object

## Subsetting both Rows and Columns

In [14]:
subset = df.loc[:,['year','pop']] #gives us the data of specified rows and columns from the dataframe

subset.head()

year       pop
0  1952   8425333
1  1957   9240934
2  1962  10267083
3  1967  11537966
4  1972  13079460

In [15]:
df.loc[df['year'] == 1967,:] #gives us the data of a specific year from the dataframe

country continent  year  lifeExp       pop    gdpPercap
3            Afghanistan      Asia  1967   34.020  11537966   836.197138
15               Albania    Europe  1967   66.220   1984060  2760.196931
27               Algeria    Africa  1967   51.407  12760499  3246.991771
39                Angola    Africa  1967   35.985   5247469  5522.776375
51             Argentina  Americas  1967   65.634  22934225  8052.953021
...                  ...       ...   ...      ...       ...          ...
1647             Vietnam      Asia  1967   47.838  39463910   637.123289
1659  West Bank and Gaza      Asia  1967   51.631   1142636  2649.715007
1671         Yemen, Rep.      Asia  1967   36.984   6740785   862.442146
1683              Zambia    Africa  1967   47.768   3900000  1777.077318
1695            Zimbabwe    Africa  1967   53.995   4995432   569.795071

[142 rows x 6 columns]

# Tidy Data

Since we have seen the basic methods and functions of the pandas library, let us have a closer look at the main focus of this notebook, i.e. Tidy data


What does Tidy Data Mean?
It is a standard way of mapping the meaning of a dataset to its structure.
In Tidy Data:

1. Each Variable forms a column
2. Each observation forms a row.
3. Each type of observational unit forms a table. 


In [16]:
import pandas as pd   #importing pandas as pd reduces the amount of typing for the coder

In [17]:
pew = pd.read_csv('pew.csv') #reading the csv file 

In [18]:
pew.head() 

religion  <$10k  $10-20k  ...  $100-150k  >150k  Don't know/refused
0            Agnostic     27       34  ...        109     84                  96
1             Atheist     12       27  ...         59     74                  76
2            Buddhist     27       21  ...         39     53                  54
3            Catholic    418      617  ...        792    633                1489
4  Don’t know/refused     15       14  ...         17     18                 116

[5 rows x 11 columns]

In [19]:
pew_long = pd.melt(pew, id_vars = 'religion') #trimming down the dataset

pew_long.head()

religion variable  value
0            Agnostic    <$10k     27
1             Atheist    <$10k     12
2            Buddhist    <$10k     27
3            Catholic    <$10k    418
4  Don’t know/refused    <$10k     15

In [20]:
billboard = pd.read_csv('billboard.csv')

billboard.head()

year        artist                    track  time  ... wk73  wk74  wk75  wk76
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22  ...  NaN   NaN   NaN   NaN
1  2000       2Ge+her  The Hardest Part Of ...  3:15  ...  NaN   NaN   NaN   NaN
2  2000  3 Doors Down               Kryptonite  3:53  ...  NaN   NaN   NaN   NaN
3  2000  3 Doors Down                    Loser  4:24  ...  NaN   NaN   NaN   NaN
4  2000      504 Boyz            Wobble Wobble  3:35  ...  NaN   NaN   NaN   NaN

[5 rows x 81 columns]

In [21]:
billboard_melt = pd.melt(
    billboard, 
    id_vars = ['year', 'artist', 'track' ,'time','date.entered'],
    var_name= 'week',
    value_name = 'rating' )     #trimming down the multiple week columns to make only one week column using melt function

In [22]:
billboard_melt.head()

year        artist                    track  time date.entered week  rating
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0
1  2000       2Ge+her  The Hardest Part Of ...  3:15   2000-09-02  wk1    91.0
2  2000  3 Doors Down               Kryptonite  3:53   2000-04-08  wk1    81.0
3  2000  3 Doors Down                    Loser  4:24   2000-10-21  wk1    76.0
4  2000      504 Boyz            Wobble Wobble  3:35   2000-04-15  wk1    57.0

In [23]:
billboard.shape #gives the original number of rows and columns in the dataset

(317, 81)

In [24]:
billboard_melt.shape #the total number of rows and columns in the melted dataframe

(24092, 7)

In [26]:
ebola = pd.read_csv('country_timeseries.csv')

ebola.head()

Date  Day  ...  Deaths_Spain  Deaths_Mali
0    1/5/2015  289  ...           NaN          NaN
1    1/4/2015  288  ...           NaN          NaN
2    1/3/2015  287  ...           NaN          NaN
3    1/2/2015  286  ...           NaN          NaN
4  12/31/2014  284  ...           NaN          NaN

[5 rows x 18 columns]

In [29]:
ebola_long = pd.melt(ebola, id_vars=['Date','Day'])

ebola_long.head()

Date  Day      variable   value
0    1/5/2015  289  Cases_Guinea  2776.0
1    1/4/2015  288  Cases_Guinea  2775.0
2    1/3/2015  287  Cases_Guinea  2769.0
3    1/2/2015  286  Cases_Guinea     NaN
4  12/31/2014  284  Cases_Guinea  2730.0

In [30]:
'Cases_Guinea'.split('_')

['Cases', 'Guinea']

In [32]:
variable_split = ebola_long['variable'].str.split('_')

variable_split


0       [Cases, Guinea]
1       [Cases, Guinea]
2       [Cases, Guinea]
3       [Cases, Guinea]
4       [Cases, Guinea]
             ...       
1947     [Deaths, Mali]
1948     [Deaths, Mali]
1949     [Deaths, Mali]
1950     [Deaths, Mali]
1951     [Deaths, Mali]
Name: variable, Length: 1952, dtype: object

In [34]:
variable_split.str.get(0) # to get the first element of the columns

0        Cases
1        Cases
2        Cases
3        Cases
4        Cases
         ...  
1947    Deaths
1948    Deaths
1949    Deaths
1950    Deaths
1951    Deaths
Name: variable, Length: 1952, dtype: object

In [35]:
variable_split.str.get(1) #to get the names of the countries

0       Guinea
1       Guinea
2       Guinea
3       Guinea
4       Guinea
         ...  
1947      Mali
1948      Mali
1949      Mali
1950      Mali
1951      Mali
Name: variable, Length: 1952, dtype: object

In [39]:
ebola_long['stats']= variable_split.str.get(0)  # here we cleaned the data and represented the data in a much appealing manner
ebola_long['country']= variable_split.str.get(1)

ebola_long.head()

Date  Day      variable   value  stats country
0    1/5/2015  289  Cases_Guinea  2776.0  Cases  Guinea
1    1/4/2015  288  Cases_Guinea  2775.0  Cases  Guinea
2    1/3/2015  287  Cases_Guinea  2769.0  Cases  Guinea
3    1/2/2015  286  Cases_Guinea     NaN  Cases  Guinea
4  12/31/2014  284  Cases_Guinea  2730.0  Cases  Guinea

In [40]:
weather = pd.read_csv('weather.csv')

In [42]:
weather.head(10)

id  year  month element  d1    d2  ...  d26   d27  d28  d29   d30  d31
0  MX17004  2010      1    tmax NaN   NaN  ...  NaN   NaN  NaN  NaN  27.8  NaN
1  MX17004  2010      1    tmin NaN   NaN  ...  NaN   NaN  NaN  NaN  14.5  NaN
2  MX17004  2010      2    tmax NaN  27.3  ...  NaN   NaN  NaN  NaN   NaN  NaN
3  MX17004  2010      2    tmin NaN  14.4  ...  NaN   NaN  NaN  NaN   NaN  NaN
4  MX17004  2010      3    tmax NaN   NaN  ...  NaN   NaN  NaN  NaN   NaN  NaN
5  MX17004  2010      3    tmin NaN   NaN  ...  NaN   NaN  NaN  NaN   NaN  NaN
6  MX17004  2010      4    tmax NaN   NaN  ...  NaN  36.3  NaN  NaN   NaN  NaN
7  MX17004  2010      4    tmin NaN   NaN  ...  NaN  16.7  NaN  NaN   NaN  NaN
8  MX17004  2010      5    tmax NaN   NaN  ...  NaN  33.2  NaN  NaN   NaN  NaN
9  MX17004  2010      5    tmin NaN   NaN  ...  NaN  18.2  NaN  NaN   NaN  NaN

[10 rows x 35 columns]

In [45]:
weather_melt = pd.melt(
    weather,
    id_vars=['id','year','month','element'], 
    var_name='day', 
    value_name= 'temp'
    )                   #in this dataframe, we reduced the multiple unnecessary coloumns and just included a single column named day 

weather_melt

id  year  month element  day  temp
0    MX17004  2010      1    tmax   d1   NaN
1    MX17004  2010      1    tmin   d1   NaN
2    MX17004  2010      2    tmax   d1   NaN
3    MX17004  2010      2    tmin   d1   NaN
4    MX17004  2010      3    tmax   d1   NaN
..       ...   ...    ...     ...  ...   ...
677  MX17004  2010     10    tmin  d31   NaN
678  MX17004  2010     11    tmax  d31   NaN
679  MX17004  2010     11    tmin  d31   NaN
680  MX17004  2010     12    tmax  d31   NaN
681  MX17004  2010     12    tmin  d31   NaN

[682 rows x 6 columns]

In [51]:
weather_tidy = weather_melt.pivot_table(
    index=['id','year','month','day'],
    columns='element',
    values='temp'
)               #In this code, we are basically including 2 separate columns for representing max and min temperatures of each day instead of irregular data that we had earlier

weather_tidy.reset_index()

element       id  year  month  day  tmax  tmin
0        MX17004  2010      1  d30  27.8  14.5
1        MX17004  2010      2  d11  29.7  13.4
2        MX17004  2010      2   d2  27.3  14.4
3        MX17004  2010      2  d23  29.9  10.7
4        MX17004  2010      2   d3  24.1  14.4
5        MX17004  2010      3  d10  34.5  16.8
6        MX17004  2010      3  d16  31.1  17.6
7        MX17004  2010      3   d5  32.1  14.2
8        MX17004  2010      4  d27  36.3  16.7
9        MX17004  2010      5  d27  33.2  18.2
10       MX17004  2010      6  d17  28.0  17.5
11       MX17004  2010      6  d29  30.1  18.0
12       MX17004  2010      7   d3  28.6  17.5
13       MX17004  2010      7  d14  29.9  16.5
14       MX17004  2010      8  d23  26.4  15.0
15       MX17004  2010      8   d5  29.6  15.8
16       MX17004  2010      8  d29  28.0  15.3
17       MX17004  2010      8  d13  29.8  16.5
18       MX17004  2010      8  d25  29.7  15.6
19       MX17004  2010      8  d31  25.4  15.4
20       MX17004  2010      8   d8  29.0  17.3
21       MX17004  2010     10   d5  27.0  14.0
22       MX17004  2010     10  d14  29.5  13.0
23       MX17004  2010     10  d15  28.7  10.5
24       MX17004  2010     10  d28  31.2  15.0
25       MX17004  2010     10   d7  28.1  12.9
26       MX17004  2010     11   d2  31.3  16.3
27       MX17004  2010     11   d5  26.3   7.9
28       MX17004  2010     11  d27  27.7  14.2
29       MX17004  2010     11  d26  28.1  12.1
30       MX17004  2010     11   d4  27.2  12.0
31       MX17004  2010     12   d1  29.9  13.8
32       MX17004  2010     12   d6  27.8  10.5

In [52]:
billboard_melt.head()

year        artist                    track  time date.entered week  rating
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0
1  2000       2Ge+her  The Hardest Part Of ...  3:15   2000-09-02  wk1    91.0
2  2000  3 Doors Down               Kryptonite  3:53   2000-04-08  wk1    81.0
3  2000  3 Doors Down                    Loser  4:24   2000-10-21  wk1    76.0
4  2000      504 Boyz            Wobble Wobble  3:35   2000-04-15  wk1    57.0

In [65]:
billboard_melt.loc[billboard_melt['track'] == 'Loser'] # to represent a specific track called Loser

year        artist  track  time date.entered  week  rating
3      2000  3 Doors Down  Loser  4:24   2000-10-21   wk1    76.0
320    2000  3 Doors Down  Loser  4:24   2000-10-21   wk2    76.0
637    2000  3 Doors Down  Loser  4:24   2000-10-21   wk3    72.0
954    2000  3 Doors Down  Loser  4:24   2000-10-21   wk4    69.0
1271   2000  3 Doors Down  Loser  4:24   2000-10-21   wk5    67.0
...     ...           ...    ...   ...          ...   ...     ...
22510  2000  3 Doors Down  Loser  4:24   2000-10-21  wk72     NaN
22827  2000  3 Doors Down  Loser  4:24   2000-10-21  wk73     NaN
23144  2000  3 Doors Down  Loser  4:24   2000-10-21  wk74     NaN
23461  2000  3 Doors Down  Loser  4:24   2000-10-21  wk75     NaN
23778  2000  3 Doors Down  Loser  4:24   2000-10-21  wk76     NaN

[76 rows x 7 columns]

In [66]:
billboard_songs = billboard_melt[['year','artist','track','time']]

In [67]:
billboard_songs.head()

year        artist                    track  time
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22
1  2000       2Ge+her  The Hardest Part Of ...  3:15
2  2000  3 Doors Down               Kryptonite  3:53
3  2000  3 Doors Down                    Loser  4:24
4  2000      504 Boyz            Wobble Wobble  3:35

In [68]:
billboard_songs.shape

(24092, 4)

In [69]:
billboard_songs.drop_duplicates()

year            artist                    track  time
0    2000             2 Pac  Baby Don't Cry (Keep...  4:22
1    2000           2Ge+her  The Hardest Part Of ...  3:15
2    2000      3 Doors Down               Kryptonite  3:53
3    2000      3 Doors Down                    Loser  4:24
4    2000          504 Boyz            Wobble Wobble  3:35
..    ...               ...                      ...   ...
312  2000       Yankee Grey     Another Nine Minutes  3:10
313  2000  Yearwood, Trisha          Real Live Woman  3:55
314  2000   Ying Yang Twins  Whistle While You Tw...  4:19
315  2000     Zombie Nation            Kernkraft 400  3:30
316  2000   matchbox twenty                     Bent  4:12

[317 rows x 4 columns]

### We can observe that, after dropping the duplicates, we just have 317 rows in the entire dataset from the 24092 initially

In [70]:
billboard_songs["id"] = range(len(billboard_songs))

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [71]:
billboards_songs.head()

year        artist                    track  time  id
0  2000         2 Pac  Baby Don't Cry (Keep...  4:22   0
1  2000       2Ge+her  The Hardest Part Of ...  3:15   1
2  2000  3 Doors Down               Kryptonite  3:53   2
3  2000  3 Doors Down                    Loser  4:24   3
4  2000      504 Boyz            Wobble Wobble  3:35   4

In [72]:
billboard_songs.to_csv('billboard_songs.csv',index=False)

In [73]:
billboard_ratings = billboard_melt.merge(
    billboard_songs, on=['year','artist','track','time']
)

In [74]:
billboard_ratings.head()

year artist                    track  time date.entered week  rating    id
0  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0     0
1  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0   317
2  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0   634
3  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0   951
4  2000  2 Pac  Baby Don't Cry (Keep...  4:22   2000-02-26  wk1    87.0  1268

In [76]:
billboard_ratings = billboard_ratings[['id','date.entered','week','rating']]

billboard_ratings.head()

id date.entered week  rating
0     0   2000-02-26  wk1    87.0
1   317   2000-02-26  wk1    87.0
2   634   2000-02-26  wk1    87.0
3   951   2000-02-26  wk1    87.0
4  1268   2000-02-26  wk1    87.0

In [77]:
billboard_ratings.to_csv('billboard_ratings.csv', index= False)

In [78]:
billboard_ratings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1830992 entries, 0 to 1830991
Data columns (total 4 columns):
 #   Column        Dtype  
---  ------        -----  
 0   id            int64  
 1   date.entered  object 
 2   week          object 
 3   rating        float64
dtypes: float64(1), int64(1), object(2)
memory usage: 69.8+ MB


#Conclusion:

## We have worked on various datasets and observed that tidy data is the most dependable way to clean a dataset and organize the given dataset in a readable manner. We've worked on various functions to know which command gives a better output each time so that it reduces the time of the user read the data. Tidy data also breaks the storage into smaller sizes as we clean the data further. 

### The End